### PR TITLE
feat: well-known did config

### PR DESCRIPTION
**Pull request by Dudleyneedham**
## fixes  https://github.com/KILTprotocol/ticket/issues/2132

Adding the well-known did config

## How to test:

adding testing

## Checklist:

- [ ] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    expect(
+      (credential = await createCredential(
+        await assertionSigner({ assertion: keypair.assertion, didDocument }),
+        origin,
+        didUri
+      ))
+    ).toBeTruthy()
+  }, 30_000)
+
+  it('fails to generate a well known did configuration credential due to bad origin', async () => {
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:65]: pls rename the test - from the code I understand we check that this fails if the origin parameter is not a URL, but that's impossible to understand from the test alone

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    ).rejects.toThrow()
+  }, 30_000)
+
+  it('get domain linkage presentation', async () => {
+    expect(
+      (domainLinkageCredential = await getDomainLinkagePresentation(credential))
+    ).toBeTruthy()
+  }, 30_000)
+
+  it('rejects the domain linkage as no signature', async () => {
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:81]:
```suggestion
  it('rejects if the domain linkage has no signature', async () => {
```

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        'bad origin',
+        didUri
+      )
+    ).rejects.toThrow()
+  }, 30_000)
+
+  it('get domain linkage presentation', async () => {
+    expect(
+      (domainLinkageCredential = await getDomainLinkagePresentation(credential))
+    ).toBeTruthy()
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:78]: again, truthiness is not particularly expressive

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    domainLinkageCredential.linked_dids[0].proof.signature = '0x'
+    await expect(
+      verifyDidConfigPresentation(didUri, domainLinkageCredential, origin)
+    ).rejects.toThrow()
+  }, 30_000)
+})
+
+afterAll(async () => {
+  await api.disconnect()
+})
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:102]: I don't love how the tests depend on each other, but I also see that this file would be much bigger if you'd properly isolate them. so be it.

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  } else if (!validUrl.isUri(claimContents.origin)) {
+    throw new Error('The claim contents origin is not a valid url')
+  } else {
+    origin = claimContents.origin
+  }
+
+  const credentialSubject: CredentialSubject = {
+    id: didUri,
+    origin,
+    rootHash: credential.rootHash,
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts:107]: is that in the spec? I'm surprised to see this property here even though it is not in the CType

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  }
+}
+
+async function asyncSome(
+  credentials: DomainLinkageCredential[],
+  verify: (credential: DomainLinkageCredential) => Promise<void>
+) {
+  for (const credential of credentials) {
+    await verify(credential)
+  }
```
Comment (rflechtner) [inline src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts:163]: could that not be
```suggestion
  await Promise.all(credentials.map(credential => 
    verify(credential)
    )
  )
```

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts (hunk 4 above)
Comment (Dudleyneedham) [reply src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:102]: Yes, it is merely there to test if it breaks.

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts (hunk 5 above)
Comment (Dudleyneedham) [reply src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts:107]: Unfortunately, we don't have that in the current context. I spoke with Antonio and he will make the custom context at some point.

Comment (Dudleyneedham) [reply src/wellKnownDidConfiguration/wellKnownDidConfiguration.ts]: I have removed the `rootHash` and added the rootHash as the `Id` in the document.

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts (hunk 1 above)
Comment (Dudleyneedham) [reply src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts:65]: Updated.

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        uses: actions/checkout@v3
+
+      - name: Configure Node and package manager
+        uses: actions/setup-node@v3.4.1
+        with:
+          node-version: "16"
+          cache: "yarn"
+
+      - name: yarn install and test
+        timeout-minutes: 5
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply .github/workflows/publish.yml]: Removed the publish as out of the scope of the PR. I will open another PR with the publish in it.

File: src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  }, 30_000)
+
+  it('generate a well known did configuration credential', async () => {
+    expect(
+      (credential = await createCredential(
+        await assertionSigner({ assertion: keypair.assertion, didDocument }),
+        origin,
+        didUri
+      ))
+    ).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/wellKnownDidConfiguration/wellKnownDidConfiguration.test.ts]: Updated.

**Review by rflechtner — APPROVED** [review]
Great, looks good now!

**Pull request merged**